### ISSUE TITLE
[Bug]: Greedy sampling (temperature=0.0) doesn't work, same prompts still give different outputs. (gemma-3-4it, ...)

### ISSUE THREAD

**Issue by vanbao1610** [vllm-project/vllm#15437]
### Your current environment

vllm==0.8.1
transformers=4.50.0

GPU A100 40 GB
model gemma-3-4b-it

SamplingParams(n=1, presence_penalty=0.0, frequency_penalty=0.0, repetition_penalty=1.0, temperature=0, top_p=1.0, top_k=-1, min_p=0.0, seed=42, stop=[], stop_token_ids=[], bad_words=[], include_stop_str_in_output=False, ignore_eos=False, max_tokens=1024, min_tokens=0, logprobs=None, prompt_logprobs=None, skip_special_tokens=True, spaces_between_special_tokens=True, truncate_prompt_tokens=None, guided_decoding=None, extra_args=None)

### 🐛 Describe the bug

I see that greedy sampling with temperature = 0.0 is causing different results to be generated each time even though it is the same prompt. I hope this problem can be solved. And if more specifically, which models actually support greedy sampling.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by DarkLight1337** [comment]
Please see https://docs.vllm.ai/en/latest/getting_started/examples/reproduciblity.html

**Comment by vanbao1610** [comment]
> Please see https://docs.vllm.ai/en/latest/getting_started/examples/reproduciblity.html

I init LLM with seed=42, max_num_seqs=1.

os.environ["VLLM_ENABLE_V1_MULTIPROCESSING"] = "0"
os.environ["VLLM_USE_V1"] = "0"

And generate with SamplingParams(max_token=1024,temperature=0.0) but the result of 3 generate times is still different.

**Comment by DarkLight1337** [comment]
Can you show your code?

**Comment by vanbao1610** [comment]
> Can you show your code?


First of all thanks for answering my question. Here is my code. 


**Code:**
import os
import torch
import gc
from PIL import Image
from transformers import Gemma3Processor
from vllm import LLM, SamplingParams


torch.cuda.empty_cache()
gc.collect()

model_dir = "gemma-3-4b-it"
os.environ["VLLM_ENABLE_V1_MULTIPROCESSING"] = "0"
os.environ["VLLM_USE_V1"] = "0"
llm = LLM(
    model=model_dir,
    max_num_seqs=1,
    max_model_len=8192,
    seed=42,
    gpu_memory_utilization=0.4,
)

processor = Gemma3Processor.from_pretrained(model_dir)


img_uri = "bee.jpg"
question = "Topic of image"
for i in range(3):
    messages = [
        {
            "role": "user",
            "content": [
                {"type": "image", "image": img_uri},
                {"type": "text", "text": question},
            ],
        },
    ]

    inputs = processor.apply_chat_template(
        messages,
        add_generation_prompt=True,
        tokenize=False,
    )
    prompts = [
        {
            "prompt": inputs,
            "multi_modal_data": {"image": Image.open(img_uri)},
        }
    ]
    output = llm.generate(prompts, sampling_params=SamplingParams(temperature=0.0, max_tokens=200))
    print("[{}] output = '{}'".format(i, output[0].outputs[0].text))



**Output:**

[0] output = 'Here's a breakdown of the image's topic:

**Main Subject:** The image focuses on a vibrant pink cosmos flower with a bumblebee actively collecting pollen. 

**Secondary Elements:** 

*   **Other Flowers:** There are other pink cosmos flowers and a few smaller red flowers in the background.
*   **Dried Flowers:** Several dried and faded flowers add a touch of texture and a sense of the natural cycle of blooming.
*   **Green Foliage:** Lush green leaves and stems provide a backdrop and context for the flowers.

**Overall Theme:** The image captures a moment of nature – the pollination process and the beauty of wildflowers. It's a charming scene of a bee enjoying a flower.'

[1] output = 'Here's a breakdown of the image's topic:

**Main Subject:** The image focuses on a vibrant pink cosmos flower with a bumblebee actively collecting pollen. 

**Secondary Elements:** 

*   **Other Cosmos Flowers:** There are several other pink cosmos flowers in the background, some in bloom and some starting to wilt.
*   **Green Foliage:** Lush green leaves and stems provide a natural backdrop.
*   **Red Flowers:** A few small red flowers add a pop of color.

**Overall Theme:** The image captures a moment of nature – the pollination process and the beauty of wildflowers. It's a classic scene of a bee interacting with a flower.'

[2] output = 'Here's a breakdown of the image's topic:

**Subject:** The primary subject is a close-up shot of a vibrant pink cosmos flower. 

**Key Elements:**

*   **Cosmos Flower:** The large, pink cosmos flower is the focal point, with a bee actively collecting pollen.
*   **Bee:** A fuzzy, black and yellow bumblebee is prominently featured on the flower, demonstrating pollination.
*   **Other Flowers:** There are other pink and red flowers in the background, adding to the floral scene.
*   **Green Foliage:** Lush green leaves and stems provide the backdrop and context for the flowers.

**Overall Theme:** The image captures a moment of nature, specifically the interaction between a bee and a flower, highlighting the beauty of springtime and the importance of pollinators.

Do you want me to describe any specific aspect of the image in more detail?'

**Comment by DarkLight1337** [comment]
cc @WoosukKwon

**Comment by zhaoxudong01** [comment]
in vllm/sampling_params.py 

301         if 0 < self.temperature < _MAX_TEMP:
302             logger.warning(
303                 "temperature %s is less than %s, which may cause numerical "
304                 "errors nan or inf in tensors. We have maxed it out to %s.",
305                 self.temperature, _MAX_TEMP, _MAX_TEMP)
306             self.temperature = max(self.temperature, _MAX_TEMP)

After commenting out this part, it will be work.

**Comment by vanbao1610** [comment]
> in vllm/sampling_params.py
> 
> 301 if 0 < self.temperature < _MAX_TEMP: 302 logger.warning( 303 "temperature %s is less than %s, which may cause numerical " 304 "errors nan or inf in tensors. We have maxed it out to %s.", 305 self.temperature, _MAX_TEMP, _MAX_TEMP) 306 self.temperature = max(self.temperature, _MAX_TEMP)
> 
> After commenting out this part, it will be work.

First, this code logic is just to set temperature = _MAX_TEMP. **_It doesn't work._**

I tested the above code with the following cases:


**vllm==0.8.2**
**transformers=4.50.0**
**flashinfer-python==0.2.2.post1**

-----------------------------------------
VLLM_ENABLE_V1_MULTIPROCESSING=1
VLLM_DISABLE_COMPILE_CACHE=1
VLLM_USE_V1=1

model = gemma-3-4b-it
seed=42
enforce_eager = true
**enable_prefix_caching = false**

**temperature=0.0**
max_token=1024

re-generate 5 times

=> **Same**

-----------------------------------------
VLLM_ENABLE_V1_MULTIPROCESSING=1
VLLM_DISABLE_COMPILE_CACHE=1
VLLM_USE_V1=1

model = gemma-3-4b-it
seed=42
enforce_eager = true
**enable_prefix_caching = false**

**temperature=1.0**
max_token=1024

re-generate 5 times

=> **Different**

-----------------------------------------
VLLM_ENABLE_V1_MULTIPROCESSING=1
VLLM_DISABLE_COMPILE_CACHE=1
VLLM_USE_V1=1

model = gemma-3-4b-it
seed=42
enforce_eager = false
enable_prefix_caching = false

temperature=0.0
max_token=1024

re-generate 5 times

=> Same

-----------------------------------------
VLLM_ENABLE_V1_MULTIPROCESSING=1
VLLM_DISABLE_COMPILE_CACHE=1
VLLM_USE_V1=1

model = gemma-3-4b-it
seed=42
enforce_eager = false
**enable_prefix_caching = true**

**temperature=0.0**
max_token=1024

re-generate 5 times

=> **Different**

-----------------------------------------
VLLM_ENABLE_V1_MULTIPROCESSING=0
VLLM_DISABLE_COMPILE_CACHE=1
VLLM_USE_V1=1

model = gemma-3-4b-it
seed=42
enforce_eager = false
enable_prefix_caching = true

temperature=0.0
max_token=1024

re-generate 5 times

=> Different

-----------------------------------------
VLLM_ENABLE_V1_MULTIPROCESSING=1
VLLM_DISABLE_COMPILE_CACHE=1
VLLM_USE_V1=0

model = gemma-3-4b-it
seed=42
enforce_eager = false
enable_prefix_caching = false

temperature=0.0
max_token=1024

re-generate 5 times

=> Different


-----------------------------------------

With the above results, when **temperature=0.0** (Greedy Sampling) ensures the output is not changed when **V1 is enabled** and **enable_prefix_caching=false** (**Not setting the seed** when initializing LLM **does not change the result** in this case.)

Hope the dev will see this feedback. Thank you for viewing and discussing with me.

**Comment by Zengwh02** [comment]
#

> > in vllm/sampling_params.py
> > 301 if 0 < self.temperature < _MAX_TEMP: 302 logger.warning( 303 "temperature %s is less than %s, which may cause numerical " 304 "errors nan or inf in tensors. We have maxed it out to %s.", 305 self.temperature, _MAX_TEMP, _MAX_TEMP) 306 self.temperature = max(self.temperature, _MAX_TEMP)
> > After commenting out this part, it will be work.
> 
> First, this code logic is just to set temperature = _MAX_TEMP. **_It doesn't work._**
> 
> I tested the above code with the following cases:
> 
> **vllm==0.8.2** **transformers=4.50.0** **flashinfer-python==0.2.2.post1**
> 
> VLLM_ENABLE_V1_MULTIPROCESSING=1 VLLM_DISABLE_COMPILE_CACHE=1 VLLM_USE_V1=1
> 
> model = gemma-3-4b-it seed=42 enforce_eager = true **enable_prefix_caching = false**
> 
> **temperature=0.0** max_token=1024
> 
> re-generate 5 times
> 
> => **Same**
> 
> VLLM_ENABLE_V1_MULTIPROCESSING=1 VLLM_DISABLE_COMPILE_CACHE=1 VLLM_USE_V1=1
> 
> model = gemma-3-4b-it seed=42 enforce_eager = true **enable_prefix_caching = false**
> 
> **temperature=1.0** max_token=1024
> 
> re-generate 5 times
> 
> => **Different**
> 
> VLLM_ENABLE_V1_MULTIPROCESSING=1 VLLM_DISABLE_COMPILE_CACHE=1 VLLM_USE_V1=1
> 
> model = gemma-3-4b-it seed=42 enforce_eager = false enable_prefix_caching = false
> 
> temperature=0.0 max_token=1024
> 
> re-generate 5 times
> 
> => Same
> 
> VLLM_ENABLE_V1_MULTIPROCESSING=1 VLLM_DISABLE_COMPILE_CACHE=1 VLLM_USE_V1=1
> 
> model = gemma-3-4b-it seed=42 enforce_eager = false **enable_prefix_caching = true**
> 
> **temperature=0.0** max_token=1024
> 
> re-generate 5 times
> 
> => **Different**
> 
> VLLM_ENABLE_V1_MULTIPROCESSING=0 VLLM_DISABLE_COMPILE_CACHE=1 VLLM_USE_V1=1
> 
> model = gemma-3-4b-it seed=42 enforce_eager = false enable_prefix_caching = true
> 
> temperature=0.0 max_token=1024
> 
> re-generate 5 times
> 
> => Different
> 
> VLLM_ENABLE_V1_MULTIPROCESSING=1 VLLM_DISABLE_COMPILE_CACHE=1 VLLM_USE_V1=0
> 
> model = gemma-3-4b-it seed=42 enforce_eager = false enable_prefix_caching = false
> 
> temperature=0.0 max_token=1024
> 
> re-generate 5 times
> 
> => Different
> 
> With the above results, when **temperature=0.0** (Greedy Sampling) ensures the output is not changed when **V1 is enabled** and **enable_prefix_caching=false** (**Not setting the seed** when initializing LLM **does not change the result** in this case.)
> 
> Hope the dev will see this feedback. Thank you for viewing and discussing with me.

work for me
